### PR TITLE
Bug 1862643: UPSTREAM: 96120: kubelet: Expose a simple Get-WinEvent shim on the kubelet logs endpoint

### PR DESCRIPTION
**Pull request by LorbusChris**
Provide an administrator a streaming view of event logs on Windows
machines without them having to implement a client side reader.

The kubelet API for querying the Linux journal is re-used for invoking
the Get-WinEvent cmdlet in a PowerShell.
Parameters that have no functional equivalence in Get-WinEvent are
ignored when assembling the command.

Only available to cluster admins.

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Enable event log collection on Windows worker nodes

```release-note
kubelet: Expose a simple Get-WinEvent shim on the kubelet logs endpoint
```

Upstream patch reference: [kubernetes#96120](https://github.com/kubernetes/kubernetes/pull/96120)

### CODE REVIEW

File: pkg/kubelet/kubelet_server_journal.go [hunk 1]
```diff
@@ -142,6 +143,39 @@ func (a *journalArgs) Args() []string {
 	return args
 }
 
+// windowsArgs returns the PowerShell and Get-WinEvent cmdlet arguments for the given args.
+func (a *journalArgs) windowsArgs() []string {
+	args := []string{
+		"-c",
```
Comment (LorbusChris) [inline pkg/kubelet/kubelet_server_journal.go:149]: NTS: This may additionally need to include: `-NonInteractive -ExecutionPolicy Bypass`

**Comment by LorbusChris** [comment]
Putting `// +build !windows` in `pkg/kubelet/kubelet_server_journal_linux.go` doesn't seem to work properly (https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_kubernetes/383/pull-ci-openshift-kubernetes-master-verify/1313465989928587264). I've create a stub file for darwin now.

**Comment by LorbusChris** [comment]
I've verified this works:
```
[lorbus@lenovo windows-machine-config-operator]$ oc adm node-logs -l kubernetes.io/os=windows --path=journal -u docker -u Microsoft-Windows-LoadPerf --since "2020-10-08 15:00:07" --until "2020-10-08 15:06:00" --tail 6


   ProviderName: docker

TimeCreated          Id LevelDisplayName Message                             
-----------          -- ---------------- -------                             
10/8/2020 3:00:13 PM  1 Information      Daemon has completed initialization 
10/8/2020 3:00:13 PM  1 Information      API listen on //./pipe/docker_engine


   ProviderName: Microsoft-Windows-LoadPerf

TimeCreated            Id LevelDisplayName Message                                                                     
-----------            -- ---------------- -------                                                                     
10/8/2020 3:04:24 PM 1001 Information      Performance counters for the WmiApRpl (WmiApRpl) service were removed       
                                           successfully. The Record Data contains the new values of the system Last    
                                           Counter and Last Help registry entries.                                     
10/8/2020 3:04:24 PM 1000 Information      Performance counters for the WmiApRpl (WmiApRpl) service were loaded        
                                           successfully. The Record Data in the data section contains the new index    
                                           values assigned to this service.                                            
10/8/2020 3:05:07 PM 1001 Information      Performance counters for the WmiApRpl (WmiApRpl) service were removed       
                                           successfully. The Record Data contains the new values of the system Last    
                                           Counter and Last Help registry entries.                                     
10/8/2020 3:05:07 PM 1000 Information      Performance counters for the WmiApRpl (WmiApRpl) service were loaded        
                                           successfully. The Record Data in the data section contains the new index    
                                           values assigned to this service.                                            


```

There are a few gotchas:
- `--since` and `--until` only support absolute times, not relative ones.
- the filtering done with `--grep` happens _after_ getting the specified max number of events from the log with `--tail`, i.e. one may end up with a short tail than expected in that case
- `--format` is not supported
- `--case-sensitive` is not supported
-  only `--boot 0` is supported really, but if boot != 0, the command will return true and not return any logs (this is due to the fact that the command is invoked twice if the boot flag is unspecified, once for boot=0 and once for boot=-1. I wanted to make it succeed and not error in that case.

PTAL @aravindhp

File: pkg/kubelet/kubelet_server_journal_test.go [hunk 2]
```diff
@@ -11,16 +12,28 @@ import (
...
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if got := tt.args.Args(); !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("journalArgs.Args() = %v, want %v", got, tt.want)
+			_, got := getLoggingCmd(&tt.args, 0)
+			if runtime.GOOS != "windows" {
+				if !reflect.DeepEqual(got, tt.wantLinux) {
+					t.Errorf("getLoggingCmd(journalArgs{}, 0) = %v, want %v", got, tt.wantLinux)
+				}
+			} else {
```
Comment (LorbusChris) [inline pkg/kubelet/kubelet_server_journal_test.go:33]: needs update for linux vs darwin

File: pkg/kubelet/kubelet_server_journal_linux.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
+// +build linux
```
Comment (LorbusChris) [inline pkg/kubelet/kubelet_server_journal_linux.go:1]: CI verify didn't like it when I used `!windows` here, which is why I had to create the darwin stub-implementation of the function

**Comment by sjenning** [comment]
We aren't going to carry something like this.  This needs to go upstream.  Unless there is some reason of which I am not aware.

**Comment by LorbusChris** [comment]
@sjenning this PR only extends an already downstream-only feature/API to make it usable on Windows nodes. `pkg/kubelet/kubelet_server_journal.go ` does not exist upstream at all.

I guess upstreaming it would be good, but then https://github.com/openshift/kubernetes/commit/1f71fb76e476b8aeb006bd4efea2420ca01c6c49 also has to be upstreamed.

**Comment by sjenning** [comment]
Wow, ok, I had no idea we carried that :grimacing:  Looks like we've been carrying it since 4.0. https://github.com/openshift/origin/pull/21701

/hold cancel
/assign @smarterclayton @sttts

File: pkg/kubelet/kubelet_server_journal.go [hunk 4]
```diff
@@ -112,36 +112,6 @@ func newJournalArgsFromURL(query url.Values) (*journalArgs, error) {
...
-			args = append(args, "--unit="+unit)
-		}
-	}
-	if len(a.Pattern) > 0 {
-		args = append(args, "--grep="+a.Pattern)
-		args = append(args, fmt.Sprintf("--case-sensitive=%t", a.CaseSensitive))
-	}
-	return args
-}
-
```
Comment (sttts) [inline pkg/kubelet/kubelet_server_journal.go:144]: please minimize the patch. The removal is cosmetics and has a chance to conflict on rebases.

File: pkg/kubelet/kubelet_server_journal_darwin.go [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// +build darwin
```
Comment (sttts) [inline pkg/kubelet/kubelet_server_journal_darwin.go:1]: call the file `patch_....go` to make clear that this comes from a carry patch.

File: pkg/kubelet/kubelet_server_journal_linux.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	}
+	if len(a.Pattern) > 0 {
+		args = append(args, "--grep="+a.Pattern)
+		args = append(args, fmt.Sprintf("--case-sensitive=%t", a.CaseSensitive))
+	}
+
+	args = append(args, "--boot", fmt.Sprintf("%d", boot))
+
+	return "journalctl", args
+}
```
Comment (sttts) [inline pkg/kubelet/kubelet_server_journal_linux.go:40]: looks like the ld func plus one line? Can't we call the old one and just add the new args?

**Comment by sttts** [comment]
Why is this not upstreamed?

**Comment by aravindhp** [comment]
> Why is this not upstreamed?

@sttts https://github.com/openshift/kubernetes/pull/383#issuecomment-713263482

**Comment by sttts** [comment]
> @sttts [#383 (comment)](https://github.com/openshift/kubernetes/pull/383#issuecomment-713263482)

Then please upstream the other change too first. It gives context to our carries and visibility.

**Comment by aravindhp** [comment]
> Then please upstream the other change too first. It gives context to our carries and visibility. Even if it is not accepted or merged upstream yet, the PR number will be in the commit message and it serves as documentation where the change is going.

@sttts Ideally I would prefer the author of https://github.com/openshift/origin/pull/21701 to open the upstream PR given they would be better suited to handle the questions that come up. But the author is @smarterclayton so not sure how quickly that will happen and we need this for the Windows release :-)

File: pkg/kubelet/kubelet_server_journal.go (hunk 4 above)
Comment (LorbusChris) [reply pkg/kubelet/kubelet_server_journal.go:144]: I don't see this as merely cosmetical, as this func is linux-only, assembling the args for the journalctl cmd. That's why I moved it into a file that only gets build on Linux, with other implementations of the same function for other OSes in other files (i.e the right implementation is chosen at compile time).

Another option that also works is to switch on  `runtime.GOOS`, which makes the diff smaller. 

Please elaborate

File: pkg/kubelet/kubelet_server_journal_darwin.go (hunk 5 above)
Comment (LorbusChris) [reply pkg/kubelet/kubelet_server_journal_darwin.go:1]: Should I rename the already-existing `kubelet_server_journal.go` as well? It stems from an earlier carry patch (https://github.com/openshift/origin/pull/21701)

File: pkg/kubelet/kubelet_server_journal_linux.go (hunk 6 above)
Comment (LorbusChris) [reply pkg/kubelet/kubelet_server_journal_linux.go:40]: Yes, I've moved the function here as it's linux-only. Implementations for other OSes return a different command with different args.

**Comment by LorbusChris** [comment]
I can work on upstreaming these commits in November. Should I open one or two seperate PRs, for the original changes from https://github.com/openshift/origin/pull/21701 and changes from this PR here?

**Comment by sttts** [comment]
Would combine them into one PR, maybe even on commit.

You can just open it as WIP PR upstream, and we can reference it in o/k. I really don't like to keep updating patches in rebase witohut the PR to be open.

**Comment by LorbusChris** [comment]
@sttts @sjenning I've opened the upstream PR: https://github.com/kubernetes/kubernetes/pull/96120/commits

File: pkg/kubelet/kubelet_server_journal.go (hunk 4 above)
Comment (sttts) [reply pkg/kubelet/kubelet_server_journal.go:144]: I compiles on every platform, doesn't it? That's my definition of cosmetical here.

Comment (sttts) [reply pkg/kubelet/kubelet_server_journal.go:144]: ic, getLoggingCmd on Windows is completely different.

**Comment by rphillips** [comment]
Upstream patch reference: https://github.com/kubernetes/kubernetes/pull/96120

**Comment by derekwaynecarr** [comment]
@aravindhp why is this required upstream?  a change like this upstream really requires a kep.

**Comment by sttts** [comment]
The upstream PR won't make it into 1.21, and probably needs a KEP according to @derekwaynecarr.

Approving as a temporary carry.

/approve

**Comment by LorbusChris** [comment]
@derekwaynecarr we need a way to get the Container Runtime logs (Docker at this time) without SSH'into the Windows machine. Docker on Windows always logs to the WinEvent log - it's not configurable.

Please note that the client side to make this feature usable should also be upstreamed to `kubectl`:
https://github.com/openshift/origin/pull/21701/commits/d9dc6892d395c239b19831641218ca047e74ad94

**Pull request merged**